# **Question 2.** Multilinear Algebra

<font size=4>
Given 

\begin{equation}
A * B = \begin{pmatrix} 1 & 2\\ 3 & 4\end{pmatrix}
, A * C = \begin{pmatrix} 5 & 6\\ 7 & 8\end{pmatrix}
,C * D + D = \begin{pmatrix} 6 & 4\\ 16 & 10\end{pmatrix}
,y = \begin{pmatrix} 1 & 2 & 3 & 4 \end{pmatrix}^T
\end{equation}

find the vector of coefficients $\hat{\beta}$ by solving the following optimization problem:

\begin{equation}
\hat{\beta} = argmin_{\beta}\|y-\{[(A \otimes C)^T * (B^T \otimes A^T)]\ [(B \odot C)* (A \odot D) + A * B \odot D]\}\beta \|^2_2
\end{equation}

Simplify the above expression to an appropriate form before solving the optimization problem.

Hint: $(A \otimes B) * (C \otimes D) = (A * C) \otimes (B * D); (A \odot B) * (C \odot D) = (A * C) \odot (B * D)$

In [1]:
import numpy as np
from tensorly.tenalg import khatri_rao
from IPython.core.display import display, HTML
display(HTML("<style>.container { width:100% !important; }</style>"))

# ***Note: Following section is for notes and references. The answer is below this section***

# Hadamard Product

<font size=5>
Element wise matrix multiplication
E.g.

\begin{equation}
A= \left( \begin{array}{ccc}
1 & 2\\
3 & 4\end{array} \right)\qquad
B= \left( \begin{array}{ccc}
5 & 6\\
7 & 8\end{array} \right)
\end{equation}

\begin{equation}
A * B = 
\left( \begin{array}{ccc}
1*5 & 2*6\\
3*7 & 4*8\end{array} \right)\qquad
= \left( \begin{array}{ccc}
5 & 12\\
21 & 32\end{array} \right)
\end{equation}

**Code**
```
import numpy as np

A = np.array([[1,2],[3,4]])
B = np.array([[5,6],[7,8]])

ham = A*B
```

# Kronecker Product
------------------------
<font size=5>
Denoted $A \otimes B$ where A $\in \mathbb{R}^{IxJ}$ and B $\in \mathbb{R}^{KxL}$. The result is a matrix of size (IK) x (JL) and defined by:

\begin{equation}
\begin{pmatrix}
a_{11}B & a_{12}B & ... & a_{1J}B \\
a_{21}B & a_{22}B & ... & a_{2J}B \\
... & ... & ... & ... \\
a_{I1}B & a_{I2}B &  ... & a_{IJ}B \\
\end{pmatrix}
\end{equation}


\begin{equation}
u \otimes v = 
\begin{pmatrix}
1*v & 2*v\\
3*v & 4*v
\end{pmatrix}
= 
\begin{pmatrix}
5 & 6 & 10 & 12\\
7 & 8 & 14 & 16\\
15 & 18 & 20 & 24\\
21 & 24 & 28 & 32\end{pmatrix}
\end{equation}

**Code**
```
import numpy as np

A = np.array([[1,2],[3,4]])
B = np.array([[5,6],[7,8]])

kron = np.kron(A,B)
```

# Khati-Rao Product
------------------------
<font size=5>
"Matching columnwise" Kronecker product. Denoted $A \odot B$ where A $\in \mathbb{R}^{IxK}$ and B $\in \mathbb{R}^{JxK}$. A $\odot$ B is a matrix of size (IJ) x (K) and computed by

\begin{equation}
A \odot B = [a_1 \otimes b_1\ \ \ \ \ a_2 \otimes b_2\ \ ....\ \  a_k \otimes b_k ]
\end{equation}

\begin{equation}
A \odot B = 
\left( \begin{array}{ccc}
1*5 & 2*6\\
1*7 & 2*8\\
3*5 & 4*6\\
3*7 & 4*8
\end{array} \right)\qquad
= \left( \begin{array}{ccc}
5 & 12\\
7 & 16\\
15 & 24\\
21 & 32
\end{array} \right)
\end{equation}


**Code**
```
from tensorly.tenalg import khatri_rao

A = np.array([[1,2],[3,4]])
B = np.array([[5,6],[7,8]])

kr = khatri_rao([A,B])
```

# **Answer:**
---------------
<font size=5>
A few things to note first

\begin{equation}I.  \ \ \ \ **A*B** \in \mathbb{R}^{2x2} , \ \ \ \ **A*C** \in \mathbb{R}^{2x2} \ \ \ \ \therefore A \in \mathbb{R}^{2x2} ,\ B \in \mathbb{R}^{2x2}\ , and\ \ C \in \mathbb{R}^{2x2}\end{equation}

\begin{equation}II. \ \ \ \ (A \otimes B)^T = (A^T\otimes B^T)\end{equation}

\begin{equation}III. \ \ \ \ (A * C) = (C * A)\ -\ commutative\ property\ holds\ for\ Hadamard\ product\end{equation}

\begin{equation}IV. \ \ \ \ (A \odot B) + (A \odot C) = A \odot (B+C)\ -\ distributive\ property\ holds\ for\ Khatri\ Rao\ product\end{equation}

<br><br><br><br><br>

\begin{equation}
\hat{\beta} = argmin_{\beta}\|y-\{[(A \otimes C)^T * (B^T \otimes A^T)]\ [(B \odot C)* (A \odot D) + A * B \odot D]\}\beta \|^2_2 \ \ \ \ \ \ \ (1)
\end{equation}

<br>

Rearraning the elements in the 1st square bracket of (1) according to item (II.) above and applying the hint $(A \otimes B) * (C \otimes D) = (A * C) \otimes (B * D)$

<br>


\begin{equation}
\hat{\beta} = argmin_{\beta}\|y-\{[(A^T \otimes C^T) * (B^T \otimes A^T)]\ [(B \odot C)* (A \odot D) + A * B \odot D]\}\beta \|^2_2
\\
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ = argmin_{\beta}\|y-\{[(A^T * B^T) \otimes (C^T * A^T)]\ [(B \odot C)* (A \odot D) + A * B \odot D]\}\beta \|^2_2 \ \ \ \ \ \ \ (2)
\end{equation}

<br>

Following the commutative property for Hadamard product as noted above in (III.) and reapplying item (II.) equation (2) can be written as

<br>

\begin{equation}
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ = argmin_{\beta}\|y-\{[(A * B)^T \otimes (A * C)^T]\ [(B \odot C)* (A \odot D) + A * B \odot D]\}\beta \|^2_2 \ \ \ \ \ \ \ (3)
\end{equation}

<br>

Rearanging elements in the second square bracket of (3) according to the hint $(A \odot B) * (C \odot D) = (A * C) \odot (B * D)$

<br>

\begin{equation}
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ = argmin_{\beta}\|y-\{[(A * B)^T \otimes (A * C)^T]\ [(B * A) \odot (C * D) + A * B \odot D]\}\beta \|^2_2 \ \ \ \ \ \ \ (4)
\end{equation}

<br>

Notating $(A * B) = (B * A) = \gamma$ (4) becomes

<br>

\begin{equation}
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ = argmin_{\beta}\|y-\{[(A * B)^T \otimes (A * C)^T]\ [\gamma \odot (C * D) + \gamma \odot D]\}\beta \|^2_2 \ \ \ \ \ \ \ (5)
\end{equation}

<br>

Notating item (IV.) above and factoring out $\gamma$

<br>

\begin{equation}
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ = argmin_{\beta}\|y-\{[(A * B)^T \otimes (A * C)^T]\ [\gamma \odot ((C * D) + D)\ ]\}\beta \|^2_2 \ \ \ \ \ \ \ (6)
\end{equation}

<br>

Replacing $\gamma$ with $\gamma = (A * B)$ in (6) we arrive at

<br>

\begin{equation}
\hat{\beta} = argmin_{\beta}\|y-\{[(A * B)^T \otimes (A * C)^T]\ [(A * B) \odot ((C * D) + D)\ ]\}\beta \|^2_2 \ \ \ \ \ \ \ (7)
\end{equation}

<br>

\begin{equation}
\boxed{\hat{\beta} = argmin_{\beta}\|y-X\beta \|^2_2 \ \ \ \ \ \therefore X = \{[(A * B)^T \otimes (A * C)^T]\ [(A * B) \odot ((C * D) + D)\ ]\}}  \ \ \ \ \ \ \ (8)
\end{equation}

In [2]:
ab = np.array([[1,2],[3,4]])
ac = np.array([[5,6],[7,8]])
cdd = np.array([[6,4],[16,10]])
y = np.array([1,2,3,4])

### Using eqn. (8) above, create **X** matrix

In [3]:
#left square bracket - Kronecker product
M1 = np.kron(ab.T, ac.T)

In [4]:
#right square bracket - khatri rao prodcut
M2 = khatri_rao([ab, cdd])

In [5]:
# X - matrix
X = M1.dot(M2)

### Using numpy to solve OLS for beta

In [6]:
betas_np = np.linalg.lstsq(X,y, rcond=None)[0]
betas_np

array([-0.0309884 ,  0.03603101])

### Manual calculation of OLS for sanity check $\hat{\beta} = (X^T X)^{-1} X^T y$

In [7]:
betas_ols = np.linalg.pinv(X.T.dot(X)).dot(X.T.dot(y))
betas_ols

array([-0.0309884 ,  0.03603101])

<font size=5>
\begin{equation}
\boxed{\hat{\beta} = \left(
\begin{array}{ccc}
-0.0309884\\
0.03603101
\end{array}
\right)}
\end{equation}